File: objc/objcfind/Sources/objcfind/FileTypes.m
#import "FindConfig.h"
#import "FileTypes.h"
#import "FileUtil.h"

@interface FileTypes ()


@property NSString *archive;
@property NSString *audio;
@property NSString *binary;
@property NSString *code;
@property NSString *font;
@property NSString *image;
@property NSString *text;
@property NSString *unknown;
@property NSString *video;
@property NSString *xml;

@property sqlite3 *db;
@property NSMutableDictionary<NSString*,NSNumber*> *extTypeCache;
@property NSMutableDictionary<NSString*,NSNumber*> *nameTypeCache;
@property BOOL nameTypeCacheLoaded;

@end

@implementation FileTypes

- (instancetype) init {
    self = [super init];
    if (self) {
        self.archive = [NSString stringWithUTF8String:T_ARCHIVE];
        self.audio = [NSString stringWithUTF8String:T_AUDIO];
        self.binary = [NSString stringWithUTF8String:T_BINARY];
        self.code = [NSString stringWithUTF8String:T_CODE];
        self.font = [NSString stringWithUTF8String:T_FONT];
        self.image = [NSString stringWithUTF8String:T_IMAGE];
        self.text = [NSString stringWithUTF8String:T_TEXT];
        self.unknown = [NSString stringWithUTF8String:T_UNKNOWN];
        self.video = [NSString stringWithUTF8String:T_VIDEO];
        self.xml = [NSString stringWithUTF8String:T_XML];

        NSString *xfindDbPath = getXfindDbPath();
        sqlite3 *db;
        int rc = sqlite3_open_v2([xfindDbPath UTF8String], &db, SQLITE_OPEN_READONLY, nil);
        if (rc == SQLITE_OK && db != nil) {
            self.db = db;
        }
        self.extTypeCache = [[NSMutableDictionary alloc] initWithCapacity:16];
        self.nameTypeCache = [[NSMutableDictionary alloc] initWithCapacity:32];
        self.nameTypeCacheLoaded = false;
    }
    return self;
}

+ (FileType) fromName:(NSString*)typeName {
    NSString *lname = [typeName lowercaseString];
    if (lname == [NSString stringWithUTF8String:T_ARCHIVE]) {
        return FileTypeArchive;
    }
    if (lname == [NSString stringWithUTF8String:T_AUDIO]) {
        return FileTypeAudio;
    }
    if (lname == [NSString stringWithUTF8String:T_BINARY]) {
        return FileTypeBinary;
    }
    if (lname == [NSString stringWithUTF8String:T_CODE]) {
        return FileTypeCode;
    }
    if (lname == [NSString stringWithUTF8String:T_FONT]) {
        return FileTypeFont;
    }
    if (lname == [NSString stringWithUTF8String:T_IMAGE]) {
        return FileTypeImage;
    }
    if (lname == [NSString stringWithUTF8String:T_TEXT]) {
        return FileTypeText;
    }
    if (lname == [NSString stringWithUTF8String:T_VIDEO]) {
        return FileTypeVideo;
    }
    if (lname == [NSString stringWithUTF8String:T_XML]) {
        return FileTypeXml;
    }
    return FileTypeUnknown;
}

+ (NSString*) toName:(FileType)fileType {
    if (fileType == FileTypeArchive) {
        return [NSString stringWithUTF8String:T_ARCHIVE];
    }
    if (fileType == FileTypeAudio) {
        return [NSString stringWithUTF8String:T_AUDIO];
    }
    if (fileType == FileTypeBinary) {
        return [NSString stringWithUTF8String:T_BINARY];
    }
    if (fileType == FileTypeCode) {
        return [NSString stringWithUTF8String:T_CODE];
    }
    if (fileType == FileTypeFont) {
        return [NSString stringWithUTF8String:T_FONT];
    }
    if (fileType == FileTypeImage) {
        return [NSString stringWithUTF8String:T_IMAGE];
    }
    if (fileType == FileTypeText) {
        return [NSString stringWithUTF8String:T_TEXT];
    }
    if (fileType == FileTypeVideo) {
        return [NSString stringWithUTF8String:T_VIDEO];
    }
    if (fileType == FileTypeXml) {
        return [NSString stringWithUTF8String:T_XML];
    }
    return [NSString stringWithUTF8String:T_UNKNOWN];
}

- (NSDictionary<NSString*,NSNumber*>*) getFileTypesForQuery:(NSString*)query andParams:(NSArray<NSString*>*)params {
    NSMutableDictionary *results = [[NSMutableDictionary alloc] init];
    sqlite3_stmt *stmt;
    sqlite3_prepare_v2(self.db, [query UTF8String], -1, &stmt, NULL);
    for (int i=0; i < [params count]; i++) {
        int rc = sqlite3_bind_text(stmt, i + 1, [[params objectAtIndex:i] UTF8String], -1, SQLITE_TRANSIENT);
        if (rc != SQLITE_OK) {
            fprintf(stderr, "error: %s\n", sqlite3_errmsg(self.db));
        }
    }
    while (sqlite3_step(stmt) == SQLITE_ROW) {
        NSString *key = [NSString stringWithUTF8String:sqlite3_column_text(stmt, 0)];
        NSNumber *fileType = [NSNumber numberWithInt:sqlite3_column_int(stmt, 1) - 2];
        [results setObject:fileType forKey:key];
    }
    sqlite3_finalize(stmt);
    return results;
}

- (void) loadNameTypeCache {
    NSString *query = @"select name, file_type_id from file_name";
    NSDictionary<NSString*,NSNumber*>* nameTypeMap = [self getFileTypesForQuery:query andParams:[[NSArray alloc] init]];
    [self.nameTypeCache addEntriesFromDictionary:nameTypeMap];
    self.nameTypeCacheLoaded = true;
}

- (FileType) getFileTypeForQuery:(NSString*)query andParams:(NSArray<NSString*>*)params {
    FileType fileType = FileTypeUnknown;
    sqlite3_stmt *stmt;
    sqlite3_prepare_v2(self.db, [query UTF8String], -1, &stmt, NULL);
    for (int i=0; i < [params count]; i++) {
        int rc = sqlite3_bind_text(stmt, i + 1, [[params objectAtIndex:i] UTF8String], -1, SQLITE_TRANSIENT);
        if (rc != SQLITE_OK) {
            fprintf(stderr, "error: %s\n", sqlite3_errmsg(self.db));
        }
    }
    if (sqlite3_step(stmt) == SQLITE_ROW) {
        fileType = sqlite3_column_int(stmt, 0) - 2;
    }
    sqlite3_finalize(stmt);
    return fileType;
}

- (FileType) getFileTypeForFileName:(NSString *)fileName {
    if (!self.nameTypeCacheLoaded) {
        [self loadNameTypeCache];
    }
    if (self.nameTypeCache[fileName]) {
        NSNumber* n = self.nameTypeCache[fileName];
        return [n intValue];
    }
//    NSString *query = @"select file_type_id from file_name where name=?";
//    NSArray<NSString*>* params = [NSArray arrayWithObject:fileName];
//    return [self getFileTypeForQuery:query andParams:params];
    return FileTypeUnknown;
}

- (FileType) getFileTypeForExtension:(NSString *)fileExt {
    if (fileExt == nil || [fileExt isEqual:@""]) {
        return FileTypeUnknown;
    }
    if (self.extTypeCache[fileExt]) {
        NSNumber* n = self.extTypeCache[fileExt];
        return [n intValue];
    }
    NSString *query = @"select file_type_id from file_extension where extension=?";
    NSArray<NSString*>* params = [NSArray arrayWithObject:fileExt];
    FileType fileType = [self getFileTypeForQuery:query andParams:params];
    [self.extTypeCache setObject:[NSNumber numberWithInt:fileType] forKey:fileExt];
    return fileType;
}

- (FileType) getFileType:(NSString*)fileName {
    if (fileName == nil || fileName.length == 0) {
        return FileTypeUnknown;
    }
    FileType fileType = [self getFileTypeForFileName:fileName];
    if (fileType != FileTypeUnknown) {
        return fileType;
    }
    NSString *ext = [FileUtil getExtension:fileName];
    return [self getFileTypeForExtension:ext];
}

- (BOOL) isArchiveFile:(NSString*)fileName {
    return [self getFileType:fileName] == FileTypeArchive;
}

- (BOOL) isAudioFile:(NSString*)fileName {
    return [self getFileType:fileName] == FileTypeAudio;
}

- (BOOL) isBinaryFile:(NSString*)fileName {
    return [self getFileType:fileName] == FileTypeBinary;
}

- (BOOL) isCodeFile:(NSString*)fileName {
    return [self getFileType:fileName] == FileTypeCode;
}

- (BOOL) isFontFile:(NSString*)fileName {
    return [self getFileType:fileName] == FileTypeFont;
}

- (BOOL) isImageFile:(NSString*)fileName {
    return [self getFileType:fileName] == FileTypeImage;
}

- (BOOL) isTextFile:(NSString*)fileName {
    FileType fileType = [self getFileType:fileName];
    return fileType == FileTypeText || fileType == FileTypeCode || fileType == FileTypeXml;
}

- (BOOL) isVideoFile:(NSString*)fileName {
    return [self getFileType:fileName] == FileTypeVideo;
}

- (BOOL) isXmlFile:(NSString*)fileName {
    return [self getFileType:fileName] == FileTypeXml;
}

- (BOOL) isUnknownFile:(NSString*)fileName {
    return [self getFileType:fileName] == FileTypeUnknown;
}

@end

File: swift/swiftfind/Sources/swiftfind/FileTypes.swift
//
//  FileTypes.swift
//  swiftfind
//
//  Created by Cary Clark on 5/12/15.
//  Copyright (c) 2015 Cary Clark. All rights reserved.
//

import Foundation
import SQLite

public enum FileType : Int {
    case unknown, archive, audio, binary, code, font, image, text, video, xml
}

public class FileTypes {
    fileprivate static let fileTypeArchiveName = "archive"
    fileprivate static let fileTypeAudioName = "audio"
    fileprivate static let fileTypeBinaryName = "binary"
    fileprivate static let fileTypeCodeName = "code"
    fileprivate static let fileTypeFontName = "font"
    fileprivate static let fileTypeImageName = "image"
    fileprivate static let fileTypeTextName = "text"
    fileprivate static let fileTypeUnknownName = "unknown"
    fileprivate static let fileTypeVideoName = "video"
    fileprivate static let fileTypeXmlName = "xml"

    private var config: FindConfig
    private var db: Connection?
    private var extTypeCache = [String: FileType]()
    private var nameTypeCache = [String: FileType]()
    private var nameTypeCacheLoaded = false

    public init() {
        config = FindConfig()
        do {
            db = try Connection(config.xfindDb, readonly: true)
        } catch {
            print("Failed to load: \(error.localizedDescription)")
            db = nil
        }
    }

    public static func fromName(_ typeName: String) -> FileType {
        let lname = typeName.lowercased()
        switch lname {
        case fileTypeArchiveName:
            return FileType.archive
        case fileTypeAudioName:
            return FileType.audio
        case fileTypeBinaryName:
            return FileType.binary
        case fileTypeCodeName:
            return FileType.code
        case fileTypeFontName:
            return FileType.font
        case fileTypeImageName:
            return FileType.image
        case fileTypeTextName:
            return FileType.text
        case fileTypeVideoName:
            return FileType.video
        case fileTypeXmlName:
            return FileType.xml
        default:
            return FileType.unknown
        }
    }

    public static func toName(_ fileType: FileType) -> String {
        switch fileType {
        case FileType.archive:
            fileTypeArchiveName
        case FileType.audio:
            fileTypeAudioName
        case FileType.binary:
            fileTypeBinaryName
        case FileType.code:
            fileTypeCodeName
        case FileType.font:
            fileTypeFontName
        case FileType.image:
            fileTypeImageName
        case FileType.text:
            fileTypeTextName
        case FileType.video:
            fileTypeVideoName
        case FileType.xml:
            fileTypeXmlName
        default:
            fileTypeUnknownName
        }
    }

    private func getFileTypesForQueryAndParams(_ query: String, _ params: [String]) -> [String: FileType] {
        var fileTypeMap = [String: FileType]()
        do {
            let stmt = try self.db!.prepare(query, params)
            for row in stmt {
                let key = row[0] as! String
                let fileTypeId = Int(row[1] as! Int64) - 1
                fileTypeMap[key] = FileType(rawValue: fileTypeId)
            }
        } catch {
            print("Failed to execute query: \(error.localizedDescription)")
        }
        return fileTypeMap
    }
    
    private func loadNameTypeCache() -> Void {
        let query = "SELECT name, file_type_id FROM file_name"
        nameTypeCache = getFileTypesForQueryAndParams(query, [])
        nameTypeCacheLoaded = true
    }

    private func getFileTypeForQueryAndParams(_ query: String, _ params: [String]) -> FileType {
        do {
            let stmt = try self.db!.prepare(query, params)
            for row in stmt {
                let fileTypeId = Int(row[0] as! Int64) - 1
                return FileType(rawValue: fileTypeId)!
            }
        } catch {
            print("Failed to execute file_name query: \(error.localizedDescription)")
        }
        return FileType.unknown
    }

    private func getFileTypeFromFileName(_ fileName: String) -> FileType {
        if (!nameTypeCacheLoaded) {
            loadNameTypeCache()
        }
        if nameTypeCache.index(forKey: fileName) != nil {
            return nameTypeCache[fileName]!
        }
//        let query = "SELECT file_type_id FROM file_name WHERE name = ?"
//        return getFileTypeForQueryAndParams(query, [fileName])
        return FileType.unknown
    }

    private func getFileTypeFromExtension(_ fileExt: String) -> FileType {
        if (fileExt.isEmpty) {
            return FileType.unknown
        }
        if extTypeCache.index(forKey: fileExt) != nil {
            return extTypeCache[fileExt]!
        }
        let query = "SELECT file_type_id FROM file_extension WHERE extension = ?"
        let fileType = getFileTypeForQueryAndParams(query, [fileExt])
        extTypeCache[fileExt] = fileType
        return fileType
    }

    public func getFileType(_ fileName: String) -> FileType {
        let fileTypeFromFileName = getFileTypeFromFileName(fileName)
        if fileTypeFromFileName != FileType.unknown {
            return fileTypeFromFileName
        }
        let fileTypeFromExtension = getFileTypeFromExtension(FileUtil.getExtension(fileName))
        return fileTypeFromExtension
    }

    public func isArchiveFile(_ fileName: String) -> Bool {
        getFileType(fileName) == FileType.archive
    }

    public func isAudioFile(_ fileName: String) -> Bool {
        getFileType(fileName) == FileType.audio
    }

    public func isBinaryFile(_ fileName: String) -> Bool {
        getFileType(fileName) == FileType.binary
    }

    public func isCodeFile(_ fileName: String) -> Bool {
        getFileType(fileName) == FileType.code
    }

    public func isFontFile(_ fileName: String) -> Bool {
        getFileType(fileName) == FileType.font
    }

    public func isImageFile(_ fileName: String) -> Bool {
        getFileType(fileName) == FileType.image
    }

    public func isTextFile(_ fileName: String) -> Bool {
        let fileType = getFileType(fileName)
        return (fileType == FileType.text || fileType == FileType.code || fileType == FileType.xml)
    }

    public func isVideoFile(_ fileName: String) -> Bool {
        getFileType(fileName) == FileType.video
    }

    public func isUnknownFile(_ fileName: String) -> Bool {
        getFileType(fileName) == FileType.unknown
    }

    public func isXmlFile(_ fileName: String) -> Bool {
        getFileType(fileName) == FileType.xml
    }
}
